### DISCUSSION TITLE
refactor spawning logic.

### DISCUSSION THREAD

**Discussion by leonjza** [sensepost/objection#527] [General]
Right now, the objection spawning logic is not great. We can't reliably do early instrumentation, and with Frida 15, some new API's are around that we can use to improve application selection etc. Some quick thoughts on what needs implementing:

- [ ] refactor the patcher to just enable debug mode in apps to use jailed testing as described here: https://www.nowsecure.com/blog/2020/01/02/how-to-conduct-jailed-testing-with-frida/
- [x] refactor the spawning logic (and the way commands work) to allow for _real_ early instrumentation
- [x] upgrade to frida 15. explore the portal feature to see what we could leverage here. https://frida.re/news/2021/07/16/frida-15-0-released/
- [ ] investigate `Java.backtrace` (https://github.com/frida/frida-java-bridge/commit/f0c4daf0d447f1139079ea0a72c36d3287323e0e)

**Comment by duraki** [comment]
any progress ? 😢

**Comment by leonjza** [comment]
Finally made some progress here! Code is not in yet, but the spawning logic rewrite happened yesterday. https://twitter.com/leonjza/status/1452516278451773448

**Comment by duraki** [comment]
Muchos grazias @leonjza for update 💯 Let us know when you push to remote, we might be of help

**Comment by leonjza** [comment]
Just pushed the first WIP for the refactor. Most things work. Theres many breaking changes too, but feel free to experiment :P Most notably, the `explore` command is now `start`.

![image](https://user-images.githubusercontent.com/1148127/138750767-d38f3e86-665d-4547-9d6c-881c496eb4a9.png)

**Comment by duraki** [comment]
> - add the `resume` repl command to resume apps

Well, I can't thank you for this. Who knows how many times I tried typing `%resume` in Objection REPL 😹 
On a serious note: this is awesome. And at what better time then with newest Frida release, eh?

I will sure be among the first who will (s)t(r)est it out 💟 Will update personal observations here asap

_edit: attaching by $pid has been ditched as it seems._

Notable diff: [Initialization of new Agent Object via CLI](https://github.com/sensepost/objection/commit/865fb7dbe4a3150bfb57651157b0710ff075ed21#diff-52a0ba5f822dd3be844987710953ee3ab78d07a00040f68cb07c7c1cd79fd8fbR21), [Gadget ditch](https://github.com/sensepost/objection/commit/865fb7dbe4a3150bfb57651157b0710ff075ed21#diff-8521cea3fd6519e58b402368e7909f39d240ef9a2cb030c1606ac0c891ed2927R21), [New Agent logic](objection/utils/agent.py)

**Comment by leonjza** [comment]
> edit: attaching by $pid has been ditched as it seems.

If you provide a pid to `--name` it should [work](https://github.com/sensepost/objection/blob/865fb7dbe4a3150bfb57651157b0710ff075ed21/objection/utils/agent.py#L223) ok.
